### PR TITLE
[SPARK-48560][SS][PYTHON] Make StreamingQueryListener.spark settable

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to make StreamingQueryListener.spark settable

### Why are the changes needed?

```python
from pyspark.sql.streaming.listener import StreamingQueryListener

class MyListener(StreamingQueryListener):
  def __init__(self, spark):
    self.spark = spark

  def onQueryStarted(self, event):
    pass

  def onQueryProgress(self, event):
    pass

  def onQueryTerminated(self, event):
    pass

MyListener(spark)
```

is broken from 3.5.0 after SPARK-42941.

### Does this PR introduce _any_ user-facing change?

Yes, end users who implement `StreamingQueryListener` can add `spark` attribute in their implementation.

### How was this patch tested?

Manually tested, and added a unittest.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by WweiL** [comment]
Thanks for the fix!

**Review by WweiL — APPROVED** [review]
Pending fixing lint

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/streaming/listener.py [hunk 1]
```diff
@@ -75,6 +75,11 @@ def spark(self) -> Optional["SparkSession"]:  # type: ignore[name-defined] # noq
         else:
             return None
 
+    @spark.setter
+    def spark(self, session):
+        # For backward compatibility
+        self._sparkSession = session
```
Comment (zsxwing) [inline python/pyspark/sql/streaming/listener.py:81]: Do we need to update `_set_spark_session` to not override if the user has already set a spark session?

**Review by xinrong-meng — APPROVED** [review]

File: python/pyspark/sql/streaming/listener.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/streaming/listener.py:81]: SGTM

**Comment by HyukjinKwon** [comment]
Will fix up the test

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by rangadi** [comment]
This is awesome. Thanks @HyukjinKwon .

**Pull request closed without merging**